confd/pkg/backends/calico/routes.go

```diff
@@ -307,6 +307,9 @@ func (rg *routeGenerator) setRoutesForKey(key string, routes []string) {
 		// Advertise route if not already advertised
 		if _, ok := advertisedRoutes[route]; !ok {
 			rg.advertiseRoute(key, route)
+			// Ensure route is added after config change
```

Comment:
Starting with an answer to just this one, because that defines what we need to do further.

I hit this bug while testing the other code, when switching between /32 and a /29.

To reproduce:
1) Add network definition in the global config for load balancer IPs
2) Create a service Local traffic policy (the IP will then get announced and be added to the  advertisedRoutes map above.
3) Change the network subnet mask in the global config, so the IP announces in the step above is still valid.
This will cause the static route (/32) to be removed, but since it still exists in the advertisedRoutes it will not be reannounced when TriggerResync() is called.

To make this /32 being advertised again you'll have to trigger a change to the service it forcing it to be withdrawn and then revert that again.
I didn't know it then, but this bug is probably the one which lead me to use /32 at the first place a couple of years ago.

Reply:
Sorry for the long delay on this - I'd still like to get this in for v3.24, though. 

Thanks for the explanation - I believe I see the issue now. Essentially, we have two different route "producers" - Services, and BGPConfiguration.

If they both produce the same route, and then one of them withdraws the route, we'll incorrectly stop advertising that route even though it still should be.

To handle cases like this, I think the right tool to use is reference counting.

I would probably do something like this.

1. Add a new `programmedRouteRefCount map[string]int` field to the client struct.
2. In `addRoutesLockHeld`, increment the ref count for each key whenever a route is added.
3. In `deleteRoutesLockHeld`, decrement the ref count whenever a route is removed, and only delete from the cache when the refcount hits 0. 

From a skim of the code, this is all I would expect us to need to do.  I've put together a PR to solve this particular case: https://github.com/projectcalico/calico/pull/6416

Let me know what you think and then it should be a simple matter of rebasing your PR on that one.

Reply:
Sorry for a bit of delay here too (vacation). I think all your comments should have been resolved by now, but I've not yet tested it in combination with your changes. The testing environment is running an older k8s version (1.20.x) and somehow the latest calico fails to start.

Didn't have time to dig into into, but I think it might be related to PodDisruptionBudget using policy/v1 which is still in beta at 1.20.